### DISCUSSION TITLE
How do you make a native library for react-native-windows?

### DISCUSSION THREAD

**Discussion by shirakaba** [microsoft/react-native-windows#13548] [Q&A] [answered]
I've set up some boilerplate using `npx --package react-native-test-app@latest init` as recommended in [react-native-test-app](https://github.com/microsoft/react-native-test-app). Now what?

One example, [react-native-randombytes](https://github.com/mvayngrib/react-native-randombytes/tree/master/windows), is 5 years old.

Another example, [react-native-document-picker](https://github.com/react-native-documents/document-picker/tree/master/windows/ReactNativeDocumentPicker), looks more recent.

There's also [react-native-slider](https://github.com/callstack/react-native-slider/tree/main/package/windows).

And [react-native-video](https://github.com/TheWidlarzGroup/react-native-video/tree/master/windows).

They've all got a bunch of boilerplate. I'm not sure how much gets set up for you after running `npm run windows:build`, but I'm surprised that `react-native-test-app` doesn't set you up with a placeholder Hello World package to just write code into.

Some of the projects use C#, some use C++, some have mentions of UWP, some seem to have different folders for different versions of RNW. Some out there involve codegen.

In my case, I'm developing for RNW v0.73. I primarily want to bring `react-native-randombytes` up to date, because it's failing to build, but I do also want to learn how to make my own native modules in general. Interested both in the workflow for C# and C++. Especially interested in how to share C++ code with macOS.

**Comment by jonthysell** [comment]
The documentation for starting adding windows support to a native module is here: https://microsoft.github.io/react-native-windows/docs/native-modules-setup

You can choose either C++ or C# for creating your native module code.

Note that there are community modules that still carry a "legacy" RNW implementation (i.e. RN < 0.60) that is no longer compatible with current RNW. It looks like [react-native-randombytes](https://github.com/mvayngrib/react-native-randombytes/tree/master/windows) is in that boat and would need a new implementation. Note: the meat of the code looks reusable (i.e. the cryptographic random generation) but all the project/infra code around it is out of date.

**Reply by shirakaba** [reply]
Thank you! Unfortunately, I have followed those docs and run into many problems; more details here. I'll close this discussion in favour of that one (I'd forgotten I'd filed this one).

https://github.com/microsoft/react-native-windows/discussions/13560

**Comment by shirakaba** [comment, marked as answer]
Closing in favour of this discussion with more concrete details: https://github.com/microsoft/react-native-windows/discussions/13560
